File: src/main/java/net/auoeke/limitless/enchantment/EnchantingBlocks.java
package net.auoeke.limitless.enchantment;

import java.util.Random;
import net.auoeke.limitless.config.LimitlessConfiguration;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.item.ItemStack;
import net.minecraft.tag.Tag;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import org.apache.commons.lang3.mutable.MutableFloat;

@SuppressWarnings("unused")
public interface EnchantingBlocks {
    String INTERNAL_NAME = "net/auoeke/limitless/enchantment/EnchantingBlocks";

    Tag<Block> tag = new EnchantingBlockTag();

    static int calculateRequiredExperienceLevel(Random random, int slotIndex, float enchantingPower, ItemStack stack) {
        if (stack.getItem().getEnchantability() <= 0) {
            return 0;
        }

        int maxEnchantingPower = LimitlessConfiguration.instance.enchantment.enchantingBlocks.maxPower;

        if (enchantingPower > maxEnchantingPower) {
            enchantingPower = maxEnchantingPower;
        }

        float level = random.nextInt(8) + 1 + enchantingPower / 4 + random.nextInt((int) enchantingPower / 2 + 1);

        return slotIndex == 0
            ? (int) Math.max(level / 3, 1)
            : (int) (slotIndex == 1
                ? level * 2 / 3 + 1
                : Math.max(level, enchantingPower));
    }

    static float countEnchantingPower(World world, BlockPos enchantingTablePos) {
        var horizontalRadiusRange = LimitlessConfiguration.instance.enchantment.enchantingBlocks.radius.xz;
        var verticalRadiusRange = LimitlessConfiguration.instance.enchantment.enchantingBlocks.radius.y;
        var power = new MutableFloat();

        forBlockInRange(world, enchantingTablePos, horizontalRadiusRange.min, horizontalRadiusRange.max, verticalRadiusRange.min, verticalRadiusRange.max, (tableBlockState, pos, dX, dY, dZ) ->
            power.add(LimitlessConfiguration.instance.enchantment.enchantingBlocks.enchantingPower(tableBlockState.getBlock()))
        );

        return power.floatValue();
    }

    static void forBlockInRange(World world, BlockPos center, int minHorizontalRadius, int maxHorizontalRadius, int minVerticalRadius, int maxVerticalRadius, EnchantingBlockConsumer action) {
        int verticalRadius;
        int horizontalRadius;
        int end;
        int displacement;
        BlockPos blockPos;
        BlockState blockState;

        for (int k = -1; k <= 1; k += 2) {
            for (verticalRadius = minVerticalRadius; verticalRadius <= maxVerticalRadius; verticalRadius++) {
                for (horizontalRadius = minHorizontalRadius; horizontalRadius <= maxHorizontalRadius; horizontalRadius++) {
                    end = k * horizontalRadius;

                    for (int distance = -horizontalRadius; distance <= horizontalRadius; distance++) {
                        displacement = k * distance;
                        blockState = world.getBlockState(blockPos = center.add(displacement, verticalRadius, end));

                        action.accept(blockState, blockPos, displacement, verticalRadius, end);

                        if (distance != -horizontalRadius && distance != horizontalRadius) {
                            blockState = world.getBlockState(blockPos = center.add(end, verticalRadius, displacement));

                            action.accept(blockState, blockPos, end, verticalRadius, displacement);
                        }
                    }
                }
            }
        }
    }
}

File: src/net/auoeke/limitless/enchantment/EnchantingBlocks.kt
package net.auoeke.limitless.enchantment

import net.auoeke.limitless.config.Configuration
import net.minecraft.block.BlockState
import net.minecraft.item.ItemStack
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import java.util.*
import kotlin.math.max
import kotlin.math.min

object EnchantingBlocks {
    const val INTERNAL_NAME = "net/auoeke/limitless/enchantment/EnchantingBlocks"

    @Suppress("NAME_SHADOWING")
    @JvmStatic
    fun calculateRequiredExperienceLevel(random: Random, slotIndex: Int, enchantingPower: Float, stack: ItemStack): Int {
        if (stack.item.enchantability <= 0) {
            return 0
        }

        val enchantingPower = min(enchantingPower, Configuration.instance.enchantment.enchantingBlocks.maxPower.toFloat())
        val level = random.nextInt(8) + 1 + enchantingPower / 4 + random.nextInt(enchantingPower.toInt() / 2 + 1)

        return when (slotIndex) {
            0 -> max(level / 3, 1f).toInt()
            else -> when (slotIndex) {
                1 -> level * 2 / 3 + 1
                else -> max(level, enchantingPower)
            }.toInt()
        }
    }

    @JvmStatic
    fun countEnchantingPower(world: World, enchantingTablePos: BlockPos): Float {
        val config = Configuration.instance.enchantment.enchantingBlocks
        val horizontalRadiusRange = config.radius.xz
        val verticalRadiusRange = config.radius.y
        var power = 0F

        forBlockInRange(world, enchantingTablePos, horizontalRadiusRange.min, horizontalRadiusRange.max, verticalRadiusRange.min, verticalRadiusRange.max) {tableBlockState, _, _, _, _ ->
            power += config.enchantingPower(tableBlockState.block)
        }

        return power
    }

    private fun forBlockInRange(world: World, center: BlockPos, minHorizontalRadius: Int, maxHorizontalRadius: Int, minVerticalRadius: Int, maxVerticalRadius: Int, action: EnchantingBlockConsumer) {
        var verticalRadius: Int
        var horizontalRadius: Int
        var end: Int
        var displacement: Int
        var blockPos: BlockPos
        var blockState: BlockState

        for (k in -1..1 step 2) {
            verticalRadius = minVerticalRadius

            while (verticalRadius <= maxVerticalRadius) {
                horizontalRadius = minHorizontalRadius

                while (horizontalRadius <= maxHorizontalRadius) {
                    end = k * horizontalRadius

                    for (distance in -horizontalRadius..horizontalRadius) {
                        displacement = k * distance
                        blockState = world.getBlockState(center.add(displacement, verticalRadius, end).also {blockPos = it})
                        action.accept(blockState, blockPos, displacement, verticalRadius, end)

                        if (distance != -horizontalRadius && distance != horizontalRadius) {
                            blockState = world.getBlockState(center.add(end, verticalRadius, displacement).also {blockPos = it})
                            action.accept(blockState, blockPos, end, verticalRadius, displacement)
                        }
                    }

                    horizontalRadius++
                }

                verticalRadius++
            }
        }
    }
}
